### PR TITLE
[FLINK-8613] [flip6] [yarn] Return excess containers

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

Upon notification of newly allocated containers, the YarnResourceManager
will only accept as many containers as there are pending container requests.
All excess containers will be returned.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 1]
```diff
@@ -325,26 +325,43 @@ public void onContainersCompleted(List<ContainerStatus> list) {
...
-			catch (Throwable t) {
-				// failed to launch the container, will release the failed one and ask for a new one
-				log.error("Could not start TaskManager in container {},", container, t);
+			log.info(
+				"Received new container: {} - Remaining pending container requests: {}",
+				container.getId(),
+				numPendingContainerRequests);
+
+			if (numPendingContainerRequests > 0) {
+				numPendingContainerRequests = Math.max(0, numPendingContainerRequests - 1);
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: Isn't it enough to write `numPendingContainerRequests--`?
If `numPendingContainerRequests` is `0`, it won't go into this branch in the next iteration.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java [hunk 2]
```diff
@@ -325,26 +325,43 @@ public void onContainersCompleted(List<ContainerStatus> list) {
...
+				// return the excessive containers
+				log.info("Returning excess container {}.", container.getId());
 				resourceManagerClient.releaseAssignedContainer(container.getId());
-				requestYarnContainer(container.getResource(), container.getPriority());
 			}
 		}
+
+		// if we are waiting for no further containers, we can go to the
+		// regular heartbeat interval
 		if (numPendingContainerRequests <= 0) {
```
Comment (GJL) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: I think going negative should not be possible.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: true

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java]: yes indeed. However, the logic should not be affected by using `<=` as the comparison operator.

**Comment by tillrohrmann** [comment]
Thanks for the review @GJL. I addressed your comments. Merging this PR.

**Pull request closed without merging**